### ISSUE TITLE
USB serial monitor doesn't work on 9600 baud rate

### ISSUE THREAD

**Issue by irukard** [microsoft/pxt-microbit#3537]
I got use case when I need to communicate with NovaFitnes SDS011 Air Quality Sensor as well as send measured and postprocessed data to PC via serial port. Since I didn't find the way to open two serial connections in MakeCode (and BTW that would be wonderful), I've decided to use:
```
serial.redirect(
SerialPin.USB_TX,
SerialPin.P2,
BaudRate.BaudRate9600
)
```
It was possible because SDS011 has an automatic mode, which send 10 byte frame every 1 second. Baud rate of 9600 was chosen because of SDS011 configuration. And here is funny thing. On 115200 (not supported by SDS) device shows up in MakeCode Serial Monitor almost instantly. On 9600, device isn't showing up... unless I open connection with correct baud rate in another program (e.g. Putty).

I don't know if it's a bug or feature. I would really appreciate the second serial port or working 9600 baud rate in MakeCode's Serial Monitor.

**Comment by irukard** [comment]
@pelikhan, @abchatra I would really appreciate your expertise :)

**Comment by irukard** [comment]
Yea... I know it's stupid... But this works ok:
```
        serial.setBaudRate(BaudRate.BaudRate115200)
        serial.writeValue("PM25", SDS011.pm25Value())
        serial.writeValue("PM10", SDS011.pm10Value())
        serial.setBaudRate(BaudRate.BaudRate9600)
```

**Comment by pelikhan** [comment]
We always assume a baud rate of 115200 in the MakeCode client. Not quite sure how we would fix this besides with the "stupid" fix.
Are looking to display data in the MakeCode ui?

**Comment by irukard** [comment]
Yes. I need MakeCode UI graphs... Our teachers know how to use it and I would really try to avoid explaining how to grab serial data using Putty.

Correct me if I am wrong. Because SDS011 data might be read in background, and sending to PC could be in background as well, the only option I see is creating blocks for sending data with use of semaphores to avoid baud rate changes in the middle of SDS011 transmission and vice versa.

**Comment by irukard** [comment]
Does Data Streamer use the same serial?
https://github.com/microsoft/pxt-hacking-stem

**Comment by pelikhan** [comment]
Lemme see if we can give you an api to avoid serial since that is not what is used internally to send data to the client.

________________________________

**Comment by irukard** [comment]
I could potentially use a new instance of MicroBitSerial instead of using uBit.serial. However cpp extension isn't compiling in 3.1 branch right now, and I really need 3.1's serial. So I will stay with that "stupid" workaround.

**Comment by pelikhan** [comment]
You can try using ``console.log`` or ``console.logValue`` to send values to MakeCode directly without using serial.

**Comment by irukard** [comment]
I saw console.log in makecode docs, but not in microbit one. I will try.

**Comment by irukard** [comment]
> You can try using `console.log` or `console.logValue` to send values to MakeCode directly without using serial.

No I can't. It uses the same baud rate and same redirection. Basically it's the same serial. Probably that's why there are no blocks for that in editor.

**Comment by pelikhan** [comment]
Right, everything goes through serial in micro:bit and we expect 115200 on the client side. So you'll need some kind of lock when you change the baud rate.

**Comment by microbit-mark** [comment]
Also raised in https://support.microbit.org/a/tickets/46936(private)

This program works on the hardware and you can view output in eg CoolTerm, but not in the MakeCode serial view.

![image](https://user-images.githubusercontent.com/31242877/128003964-55b626f1-888f-449b-a1b8-e0125a5f4c61.png)

Is this something that can be fixed @abchatra @pelikhan

**Comment by abchatra** [comment]
@microbit-mark . Program needs serial write line to start parsing. So add the it after write value or at certain count to start the values to appear.
![image](https://user-images.githubusercontent.com/6107272/128045657-52b3b955-0b23-43e8-9cc5-d6e054ddadfc.png)

This is how the apis worked for a long time.

**Comment by microbit-mark** [comment]
@abchatra I don't think that is the issue here. The program in https://github.com/microsoft/pxt-microbit/issues/3537#issuecomment-891747055 works if the baud rate is set to 115200 (default). You can see the output in the device console

This works:

<img width="289" alt="image" src="https://user-images.githubusercontent.com/31242877/128176343-9f237a46-9a77-4d68-a722-e6aed875a36f.png">

This does not:

<img width="371" alt="image" src="https://user-images.githubusercontent.com/31242877/128176390-96001211-84dc-4e77-a5dd-1f551f458389.png">

**Comment by abchatra** [comment]
@pelikhan @mmoskal do you know more about why the baud rate makes a difference?

**Comment by abchatra** [comment]
@microbit-mark Also add a new line and see if that fixes the issue

**Comment by mmoskal** [comment]
You would need to set the interface chip to the 9600 baud rate. I don't think we ever do that, but there's a CMSIS Dap command to do so (it's just hard-coded to 115200 right now).

You can try to use external serial monitor program and specify baud rate there.

**Comment by microbit-mark** [comment]
> You can try to use external serial monitor program and specify baud rate there.

Yeah it works fine if I set the baud rate in the MakeCode program and use https://coolterm.en.lo4d.com/ to read the serial output. It just means I have to use an external tool, rather than MakeCode to monitor output.

**Comment by microbit-mark** [comment]
Re-opening this as a feature request to support additional baud rates in the serial console log

**Comment by abchatra** [comment]
@microbit-mark this is enhancement request for daplink right? Can you move this issue to the daplink repo?

**Comment by mmoskal** [comment]
Daplink allows this - you would need to add UI in MakeCode with list of baud rates. IMHO this is too complicated for our users.

**Comment by microbit-mark** [comment]
For my understanding then, Am I right in thinking that you can use the `set serial baud rate` block to change the baud rate on the hardware...

![image](https://user-images.githubusercontent.com/31242877/130218684-84b611df-fc58-4910-a901-921958208a8d.png)

...but this is not reflected in the browser serial console, as @pelikhan mentioned, you expect `115200` and it is this that would require additional UI?

If that is the case then it does seem complex. 

@abchatra I think we should at least note that the console serial log can only be used with the default baud rate. I'd expect to see this on https://makecode.microbit.org/reference/serial/set-baud-rate which doesn't appear to exist at the moment. << cc @ganicke

**Comment by Darshnik253** [comment]
I am making project smart in that i have combined bluetooth and usb serial communication code. But if i run both simultaneously then its not working(getting error 020 sad icon). so could you help in that?
